src/v/kafka/server/handlers/client_quotas.cc

```diff
@@ -99,6 +106,180 @@ values_data get_value_data(const entity_value& val) {
+    } else if (component.entity_type == "client-id-prefix") {
+        return entity_key::part{
+          .part = entity_key::client_id_prefix_match{*component.match}};
+    } else if (
+      component.entity_type == "user" || component.entity_type == "ip") {
+        return kerror{
+          error_code::unsupported_version,
+          fmt::format(
+            "Entity type '{}' not yet supported", component.entity_type),
+        };
+    } else {
+        return kerror{
+          error_code::unsupported_version,
+          fmt::format(
+            "Custom entity type '{}' not supported", component.entity_type),
+        };
+    }
+}
+
+result<entity_key::part, kerror>
+default_match_key(const component_data& component) {
+    if (component.entity_type == "client-id") {
+        return entity_key::part{.part = entity_key::client_id_default_match{}};
+    } else if (component.entity_type == "client-id-prefix") {
+        return kerror{
+          kafka::error_code::invalid_request,
+          "Invalid quota entity type, client-id-prefix entity should not "
+          "be used at the default level (use client-id default instead)."};
+    } else if (
+      component.entity_type == "user" || component.entity_type == "ip") {
+        return kerror{
+          error_code::unsupported_version,
+          fmt::format(
+            "Entity type '{}' not yet supported", component.entity_type),
+        };
+    } else {
+        return kerror{
+          error_code::unsupported_version,
+          fmt::format(
+            "Custom entity type '{}' not supported", component.entity_type),
+        };
+    }
+}
+
+using key_part_predicate = std::function<bool(const entity_key::part&)>;
+
+template<typename... Args>
+key_part_predicate make_any_filter() {
+    return [](const entity_key::part& p) {
+        return (std::holds_alternative<Args>(p.part) || ...);
+    };
+}
+
+result<key_part_predicate, kerror>
+any_match_filter(const component_data& component) {
+    if (component.entity_type == "client-id") {
+        return make_any_filter<
+          entity_key::part::client_id_default_match,
+          entity_key::part::client_id_match>();
+    } else if (component.entity_type == "client-id-prefix") {
+        return make_any_filter<entity_key::part::client_id_prefix_match>();
+    } else if (
+      component.entity_type == "user" || component.entity_type == "ip") {
+        return {
+          error_code::unsupported_version,
+          fmt::format(
+            "Entity type '{}' not yet supported", component.entity_type),
+        };
+    } else {
+        return {
+          error_code::unsupported_version,
+          fmt::format(
+            "Custom entity type '{}' not supported", component.entity_type),
+        };
+    }
+}
+
+result<entity_key::part, kerror> make_part(const auto& entity) {
+    entity_key::part part;
+    if (entity.entity_type == "client-id") {
+        if (entity.entity_name.value_or("") == "") {
```

Comment:
nit: i had to stop and think that this (and also line 204) means `!entity.entity_name.has_value()` so i would consider changing it, if you have to do a second pass on this commit

Reply:
my bad, I thought that `""` was not allowed like we don't allow it in some string properties in config_store